doc/dev/cephfs-mirroring-tree-traversal.rst

```diff
@@ -0,0 +1,176 @@
+*****************************************
+CephFS Mirroring Efficient Tree Traversal
+*****************************************
+
+Efficient traversal algorithm for finding changes between two directory trees. Let's tag the two directory trees as `source` and `sink`. I/O operations are done on the `source` directory tree. The goal is to figure out the list of inodes that "differ" between `source` and `sink`. Repoting in incremental; firgure out the list of changes by traversing the directory tree top-down. Subsequent traversal reports changes that were done _after_ the inode was checked (on previous traversal).
+
+Recursive accounting
+####################
+
+Recursive accounting of changed time is a prerequisite for algorithm effciency. Changes to a subtree triggers a recursive (bottom-up) update of a special extended attribute (`xattr`) up root. Xattr value is the time at which the I/O operation is done on an inode. This timestamp is recursively applied upto root.
```

Comment:
What's the 'cost' of propagating each modification all the way up to the root (multiple xattr updates), vs. traversing all subdirs on sync? (I assume some of the root will be cached since it'll be accessed so frequently just for the xattr updates) if we update just the subdir in which changes took place? The root may be the 'dirty bit' that can be used to indicate if we even need to do the traversal anyway, but is it really worth all the subdir io?

Reply:
Recursive accounting in CephFS already exists (recursive status) and happens lazily post I/O operation. However, I'm not sure if there were efforts done to measure the cost. @batrick @ukernel ?

Reply:
rstats for snapshot is broken for now. rstats are lazily propagated to root (there are pending patch). So for using rstats to find modification is not safe.

Reply:
right -- those fixes need to get in for this to work.

Reply:
Regardless of the fixes, I'm asking the general question - is it worthwhile the lazy attribute updates along the tree up to the root.

Reply:
> if we update just the subdir in which changes took place?

Do you mean have the client update the subdir anytime it makes a modification? The MDS is doing that already as part of the recursive stats tracking (and propagates that up to the root directory).

>  but is it really worth all the subdir io?

We're already doing this subdir I/O as part of whatever metadata mutation triggered it. The propagation up to the root is added cost but not significant when grouped with other metadata traffic.